src/ui/public/visualize/spy.js

```diff
@@ -1,106 +1,182 @@
-          setSpyMode(modeName ? null : defaultMode);
+          // If the spy panel is already shown (a currentMode is set),
+          // close the panel by setting the name to null, otherwise open the
+          // panel (i.e. set it to the default mode name).
+          if ($scope.currentMode) {
+            $scope.setSpyMode(null);
+            $scope.forceMaximized = false;
+          } else {
+            $scope.setSpyMode(defaultModeName);
+            checkForcedMaximized();
+          }
         };
 
-        $scope.toggleFullPage = function () {
-          fullPageSpy = !fullPageSpy;
-          $scope.spy.mode = getSpyObject();
+        /**
+         * Should we currently show the spy panel. True if a currentMode has been set.
+         */
+        $scope.shouldShowSpyPanel = () => {
+          return !!$scope.currentMode;
         };
 
-        $scope.onSpyModeChange = function onSpyModeChange() {
-          setSpyMode($scope.selectedModeName);
+        /**
+         * Toggle maximized state of spy panel and update the UI state.
+         */
+        $scope.toggleMaximize = function () {
+          $scope.maximizedSpy = !$scope.maximizedSpy;
+          updateUiState();
         };
 
-        if ($scope.uiState) {
-          // sync external uiState changes
-          const syncUIState = () => $scope.spy.mode = $scope.uiState.get('spy.mode');
-          $scope.uiState.on('change', syncUIState);
-          $scope.$on('$destroy', () => $scope.uiState.off('change', syncUIState));
-        }
+        /**
+         * Whenever the maximized state changes, we also need to toggle the class
+         * of the visualization.
+         */
+        $scope.$watchMulti(['maximizedSpy', 'forceMaximized'], () => {
+          $scope.visElement.toggleClass('spy-only', $scope.maximizedSpy || $scope.forceMaximized);
+        });
+
+        /**
+         * Watch for changes of the currentMode. Whenever it changes, we render
+         * the new mode into the template. Therefore we remove the previously rendered
+         * mode (if existing) and compile and bind the template of the new mode.
+         */
+        $scope.$watch('currentMode', (mode, prevMode) => {
+          if (mode === prevMode && (currentSpy && currentSpy.mode === mode)) {
+            // When the mode hasn't changed and we have already rendered it, return.
+            return;
+          }
+
+          const newMode = spyModes.byName[mode];
 
-        // re-render the spy when the name of fill modes change
-        $scope.$watchMulti([
-          'spy.mode.name',
-          'spy.mode.fill'
-        ], function (newVals, oldVals) {
-          // update the ui state, but only if it really changes
-          const changedVals = newVals.filter((val) => !_.isUndefined(val)).length > 0;
-          if (changedVals && !_.isEqual(newVals, oldVals)) {
-            if ($scope.uiState) $scope.uiState.set('spy.mode', $scope.spy.mode);
+          if (currentSpy) {
```

Comment:
when does this happen ?

Reply:
This happens every time you switch tabs in an open spy panel. In this case we destroy the open tab and render a new tab.